### PR TITLE
Haz3l checkbox and slider livelits

### PR DESCRIPTION
**Pull request by 7h3kk1d**
Open questions:

- Do we want to have parens after livelit applications in this pass that doesn't support livelit params?
- Should the rendered livelits have a span or other bounding box that fixes the width/height?
  - Currently I'm just styling the input elements

### CODE REVIEW

File: src/haz3lcore/Measured.re
Comment (7h3kk1d) [inline src/haz3lcore/Measured.re]: This is unused. Not sure where this map will live.

File: src/haz3lcore/lang/Form.re
Comment (7h3kk1d) [inline src/haz3lcore/lang/Form.re]: This allows for a zero-character livelit just so you can type `^`

File: src/haz3lcore/statics/TermBase.re
Comment (7h3kk1d) [inline src/haz3lcore/statics/TermBase.re]: Unused so far.

File: src/haz3lweb/view/Code.re
Comment (7h3kk1d) [inline src/haz3lweb/view/Code.re]: Not sure what to do about this. `inject` is a function therefore unhashable so this function doesn't get memoized. There's probably a smarter way to do the event handling I'm not thinking of. I'd rather not have to do something that reflects over the dom and captures the input elements.

File: src/haz3lcore/Measured.re
Comment (disconcision) [inline src/haz3lcore/Measured.re]: break this off into a fn pls

File: src/haz3lcore/lang/Form.re
Comment (disconcision) [reply src/haz3lcore/lang/Form.re]: So right now there's no cursor-info or error-hole feedback if you have an unbound livelit name. Eventually we'll want a legit error status for this... This isn't too hard to do now, but it's more streamlined after the ADT merge goes in (which rewrites statics and cursorinfo, so probably easiest to wait regardless).

For a workaround at the moment lets ensure that unbound livelits get highlighted in red as invalid syntax. To do this, add the current is_livelit predicate as a clause of is_bad_lit, and make is_livelit match only the two exact strings of the current two livelits.

File: src/haz3lcore/dynamics/elaborator.re
Comment (disconcision) [inline src/haz3lcore/dynamics/elaborator.re]: make a type alias for Id.Map.t(DHExp.t))... something like Livelit.state

File: src/haz3lcore/dynamics/elaborator.re
Comment (disconcision) [inline src/haz3lcore/dynamics/elaborator.re:279]: make this whole case a function in Livelit.re i think

File: src/haz3lcore/zipper/Editor.re
Comment (disconcision) [inline src/haz3lcore/zipper/Editor.re]: alias

File: src/haz3lcore/zipper/Editor.re
Comment (disconcision) [inline src/haz3lcore/zipper/Editor.re]: alias

File: src/haz3lweb/Update.re
Comment (disconcision) [inline src/haz3lweb/Update.re]: i feel these highly nested expansions are hard to read... maybe break this out into a couple appropriately-chosen lets. alternatively worth checking if someone already made getters/setters for these subcomponents, probably not worth adding if there arent any though

File: src/haz3lweb/view/BackpackView.re
Comment (disconcision) [inline src/haz3lweb/view/BackpackView.re]: alias

File: src/haz3lweb/view/Code.re
Comment (disconcision) [inline src/haz3lweb/view/Code.re]: i think in general with an appropriately named alias type we can get away with renaming livelit_state everywhere to simply 'livelits'. debatable tho your call

File: src/haz3lweb/view/Code.re
Comment (disconcision) [reply src/haz3lweb/view/Code.re]: @dmoon1221 any thoughts?

**Review by disconcision — CHANGES_REQUESTED** [review]
Pretty solid. A couple of functional issues that need addressing, some optional aesthetic stuff, and some minor code cleanup.

Functional issues: 
1. livelits seem to be missing a caret position, i.e. pressing right on "^slide|r(-------)" jumps to "^slider(-------)|". The livelit UI should be treated as a single character. Not sure how straightforward this is; I can imagine this complicating the interaction model, e.g. do we need to special-case what happens when we backspace to the right of a livelit? One general solution I can think of is to have a special terminating character on livelits which is treated as 'virtual'. Like lets say livelits required a "^" at the end (maybe instead of the start?), and that "^" is rendered in a different faded color, so it feels less weird that the caret skips over it... maybe try to make it feel stylistically more part of the livelit UI than the name. I sort of like how (I imagine) this feels... right now it feels mildly jarring when the livelit UI appears; if the appearance was triggered by pressing the special livelit character I could see that feeling quite nice. alternatively maybe the special terminating character could literally just be a space, but i think our current whitespace special-casing probably precludes that. in any case, there are some options here; the current caret movement feels a bit too janky though. @cyrus- do you have an opinion here? i think the postfix-^ is worth considering
2. unbound livelits currently trigger a (caught) elaboration exception. this should be fixed by the unbound livelit workaround suggested in the comments.

aesthetic quibbles:
Screen Shot 2023-03-16 at 12 44 35 PM
we might consider:
1. giving the UIs more room to breath by adding a half-space margin on each side
2. doing a small manual top- or bottom- margin adjustment to center them a bit better with the text
3. giving the livelit names a more UI-element-aligned color
4. maybe change the UI-element greys to browns to be more in-line with the color scheme (not pictured; debatable)
5. having some slightly punchier livelit names

all of these are marginal/debatable so don't feel obligated though i do think at least 1 is a good idea

File: src/haz3lcore/lang/Form.re
Comment (7h3kk1d) [reply src/haz3lcore/lang/Form.re]: Fixed in https://github.com/hazelgrove/hazel/pull/955/commits/0495cc50adabd9d3a7447f71eaf258dc5d8fd9bd.

File: src/haz3lcore/zipper/Editor.re
Comment (7h3kk1d) [reply src/haz3lcore/zipper/Editor.re]: This ones a value unless you were thinking of something like `Livelit.empty_livelit_state`

File: src/haz3lweb/Update.re
Comment (7h3kk1d) [reply src/haz3lweb/Update.re]: I simplified it a bit in https://github.com/hazelgrove/hazel/pull/955/commits/0abde16e81a1f11a4a076b1ba2439cdc9ba5a3e7. I was looking for a better way to do nested record updates and didn't find anything. Not sure if there's something like lenses or if I should just add more update functions.

File: src/haz3lweb/view/Code.re
Comment (7h3kk1d) [reply src/haz3lweb/view/Code.re]: I did end up refactoring this a bit in the livelit params branch https://github.com/hazelgrove/hazel/commit/2b8b9a2e2adef8575237f633b40cc275d99bef07#diff-98923b69b71582fa885895acfb8efef0a607453ecf764a5cedd4021db6521d3b. I was trying to separate out those changes but I can see if that parts easy to reason in and if it would simplify.

File: src/haz3lweb/Update.re
Comment (7h3kk1d) [reply src/haz3lweb/Update.re]: Addressed by moving https://github.com/hazelgrove/hazel/pull/955/commits/d52669ea55633a8be6a376b244cf1a088b6d03cb outside of the span. I haven't performance-tested anything in the PR though so I'm not sure how expensive the livelit rendering is.

File: src/haz3lweb/Update.re
Comment (disconcision) [inline src/haz3lweb/Update.re]: `ed => { <a>(ed) }`, could just be `<a>` bc currying

File: src/haz3lweb/view/LivelitView.re
Comment (disconcision) [inline src/haz3lweb/view/LivelitView.re]: is there a particular reason you use on_change for the checkbox but on_input for the slider? I don't totally grasp the implications of using one versus the other; i ask mostly because i'm thinking about generic ways to automatically reconcile DOM and model state in more generic contexts

Comment (7h3kk1d) [reply src/haz3lweb/view/LivelitView.re]: I had changed the range to use `on_input` so that the changes are reflected while you're sliding which is really convenient if you're trying to pick a value. I had tried it with the checkbox and had a bug, but now I think that's unrelated. I just tested both with `on_input` and it seems to work.

Addressed in https://github.com/hazelgrove/hazel/pull/955/commits/58eb82ba66e6969bffe2e710918d2f917cecf5ff

**Comment by cyrus-** [comment]
closing in favor of newer efforts in this direction based on #1218

**Pull request closed without merging**
